flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionService.java

```
@@ -0,0 +1,219 @@
+
+	KubernetesLeaderElectionService(
+			FlinkKubeClient kubeClient,
+			Executor executor,
+			KubernetesLeaderElectionConfiguration leaderConfig) {
+
+		this.kubeClient = checkNotNull(kubeClient, "Kubernetes client should not be null.");
+		this.executor = checkNotNull(executor, "Executor should not be null.");
+		this.configMapName = leaderConfig.getConfigMapName();
+		this.leaderElector = kubeClient.createLeaderElector(leaderConfig, new LeaderCallbackHandlerImpl());
+		this.leaderContender = null;
+		this.configMapLabels = KubernetesUtils.getConfigMapLabels(
+			leaderConfig.getClusterId(), LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY);
+	}
+
+	@Override
+	public void internalStart(LeaderContender contender) {
+		CompletableFuture.runAsync(leaderElector::run, executor);
+		kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+	}
+
+	@Override
+	public void internalStop() {
+		if (kubernetesWatch != null) {
+			kubernetesWatch.close();
+		}
+	}
+
+	@Override
+	protected void writeLeaderInformation() {
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (leaderElector.hasLeadership(configMap)) {
+						// Get the updated ConfigMap with new leader information
+						if (confirmedLeaderAddress != null && confirmedLeaderSessionID != null) {
+							configMap.getData().put(LEADER_ADDRESS_KEY, confirmedLeaderAddress);
+							configMap.getData().put(LEADER_SESSION_ID_KEY, confirmedLeaderSessionID.toString());
+						}
+						configMap.getLabels().putAll(configMapLabels);
+						return Optional.of(configMap);
+					}
+					return Optional.empty();
+				}).get();
+		} catch (Exception e) {
+			leaderContender.handleError(new Exception("Could not update ConfigMap " + configMapName, e));
+		}
+	}
+
+	@Override
+	protected boolean hasLeadership() {
+		return kubeClient.getConfigMap(configMapName)
+			.map(leaderElector::hasLeadership)
+			.orElse(false);
```

What is the meaning of a non-existent ConfigMap with name `configMapName`? Can this ever happen? Is this an invalid state where we should fail hard?

---

If we agree to not handle the externally deletion/update, then we should fail hardly if the ConfigMap not exists.